### PR TITLE
Organizations MVP.

### LINKED ISSUE #692: Prepare findings from user research on Organization Pages
Related to #674 and #669.

Once user research sessions are completed, compile findings from research to share with the team. Include recommendations for any revisions to the wireframes that emerge as trends from user feedback.

### LINKED ISSUE #695: Create wireframes for changes and additions needed to accommodate Organization Page workflows
Create wireframes and/or high-fidelity mock-ups for changes and additions needed to accommodate Organization Page workflows throughout the rest of the site.

I'll need to make sure this list is complete by talking more with @Coop56, but here are a few needs that I know we'll need to include in this effort:
- Pricing page for Organization Pages
- Form for initial creation of an Organization Page
- Administrative features of Organization Page
  - Adding additional admins
  - Removing admins
  - Receiving a notification that someone has added you as an admin
- Adding/removing Organization Members to Organization Page
- Editing "About" content for Organization Page
- Adding/editing/changing avatar for Organization Page
- Switching between acting as a regular user and acting as an organization
  - Adding content to the "Featured Content" area (other users' StoryLayers and MapStories)
  - Importing layers as an organization
  - Uploading icons as an organization
  - Creating MapStories as an organization
  - Publishing, editing, and saving drafts of journal entries as an organization
  - Removing or deleting/undoing any of the above

### LINKED ISSUE #910: Organization Pages
As a community organizer I want an Organization page with MapStories, Layers, Activity Feed and members so that I can showcase the Organization's contributions.

- [x] General information
- [x] Members section
- [x] Layer section
- [x] MapStories section
- [x] Activity feed section

### LINKED ISSUE #911: MapStory Admin: Create new Organization
User Story
-----------
**As a** MapStory Administrator
**I want** a new section in the admin panel for Organizations
**so that** I can create new and view navigate all Organizations on the site.

### LINKED ISSUE #912: MapStory Admin: Enable/Disable Organizations
User Story
-----------

**As a** MapStory Administrator
**I want** to remove Organization Pages from public view without deleting them
**so that** I can suspend an Organization Page for violation of any rules or norms for a period of time.

### LINKED ISSUE #913: MapStory Admin: Organization Remove
User Story
-----------

**As a** MapStory Administrator
**I want** to delete an Organization page without deleting user accounts or layer or story content from the django-admin panel
**so that** I can remove Organization Pages

### LINKED ISSUE #916: Organization management backend
User Story
-----------

**As a** Organizer (aka. Organization Manager)
**I want** a page where I can change the settings for the Organizaion I have permissions for
**so that** I can manage all things Organizations.

An Organizer can do the following:

- [ ] Add/remove other Organizers
- [ ] Add/remove Organization members
- [ ] Change the payment plan
- [ ] Edit the Organization's display information
- [ ] Upload an icon 

Mockup
--------
[Mockup needed]

### LINKED ISSUE #917: Add Story/Layer to Organization
User Story
-----------

**As an** Organizer
**I want** to see a "Add to Organization" button inside Story/Layer detail pages that are owned by members in my organization
**so that** I can quickly add or remove Layes/Stories to my Organization's page.

The Organizer can do the following when on a Story/Layer view page:

- [ ] Click a button to add the Story/Layer to the Organization
- [ ] Click the same button to remove the Story/Layer once it has been added

Mockups
---------

[Mockups needed]

### LINKED ISSUE #958: Organizations detail page
When I visit a URL an Organization's details will be displayed.

### LINKED ISSUE #959: Organization's Admin Page
A URL where the Organization's admin can manage the Organization. The user should be able to change any of the fields displayed on the Organizations Page.

### LINKED ISSUE #963: Organizations: Featured Layer Cards
A card for displaying a featured Layer inside Organization detail page.

### LINKED ISSUE #964: Organizations: Member cards
A card to represent a member of an Organization inside the Organization detail page.

### LINKED ISSUE #986: Organizations Badge Image
Display an image badge for Organizations

### LINKED ISSUE #987: Organization About
Organization About section

### LINKED ISSUE #988: Organization Page Navigation Menu
Clicking on a Link will show the corresponding section

### LINKED ISSUE #989: Organization Page Social media icons
Displays the organizations social media accounts with icons

### LINKED ISSUE #990: Organization Layer Management Buttons
Show buttons for adding Layer/Mapstory to an organization inside Layer/Mapstory view when the user has permissions.

### LINKED ISSUE #991: Organization Mapstory management button
Display a button for Mapstory management inside Organizations page.

### LINKED ISSUE #997: Url endpoints for Organization Management
URLs for Organization management.

### LINKED ISSUE #998: Organizations: Prevent the same Layer/Mapstory from being added twice
There is a bug that happens when the same layer/mapstory is added to the organization twice.

### LINKED ISSUE #1000: Handle Organization Management forms
Sanitize and process the form submitted for Organization Management.

### LINKED ISSUE #1001: Organization Image upload
Handle uploading an image for Organizations

### LINKED ISSUE #1002: Fix HTML layout.
There is a lot of  broken inherited HTML templates that are causing the Organization page to break and not display correctly. Need to find these broken tags and fix them.

### LINKED ISSUE #1003: Organization's Member management
The user can manage memberships from the Organizations Management page.

### LINKED ISSUE #1008: Add mapstory to organization button.
A button to add a Mapstory to an organizastion

### LINKED ISSUE #1009: Styles for Organizations Cards
Cards buttons and styles.

### LINKED ISSUE #1010: Featured Cards Add/Remove.
Add and remove cards from featured section if the user has the correct permisions.

### LINKED ISSUE #1026: All user messages display as warnings.
warning type is hardcoded for user messages. This comes from geonode.

### LINKED ISSUE #1086: Change Organization's IDs to slugs
Need to use slugs instead of ids for consistency.

### PR DESCRIPTION
**Pull request by josellausas**
# About

Adds new `Organizations` app to MapStory.

# New Features:

- Organization Pages can now be created by the admin.
- Multiple Users can be assigned as managers for an Organization.
- Managers can edit Organization Page.
- Managers can add/remove MapStories to an Organization Page.
- Managers can add/remove Layers to an Organization Page.
- Users can request to join an Organization.
- Managers can approve/reject membership requests. 
- Managers can feature/un-feature MapStories from the Organization's Page.
- Managers can feature/un-feature Layers from the Organization's Page .
- Managers can edit social media links for the Organization's Page.
- Managers can edit Organization's contact links.
- A user can add Layers to the Organization's page if he is a member.
- A user can add/remove MapStories to the Organization's page if he is a member.

# Issues closed by this PR

Closes #1000, closes #1001, closes #1002, closes #1003, closes #1026, closes #1009, closes #1010, closes #1008, closes #990, closes #998, closes #997, closes #959, closes #991, closes #989, closes #988, closes #987, closes #958, closes #910, closes #964, closes #917, closes #695, closes #692, closes #963, closes #916, closes #913, closes #912, closes #911, closes #1086, closes #986  

# Guides and How-To

## Create a new Organization

1. Go to the django admin (`/admin`) and login as admin.
2. Go to the Organizations section and select Organization
3. Create a new Organization. 

screen shot 2017-11-03 at 3 40 58 pm

4. Assign a user as a Organization manager (must check `is_admin`):

screen shot 2017-11-03 at 3 41 11 pm

5. You can find the new Organization at `???.com/organizations/<organization slug>` . 
6 . The admin can manage the new Organization at `???.com/organizations/<organization slug>/manager`

### CODE REVIEW

File: mapstory/apps/organizations/forms.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    LINKS & SOCIAL MEDIA
+    --------------------
+        - URLs
+        - Facebook
+        - Twitter
+        - LinkedIn
+        - Github
+        - Instragram
+    """
+    url0 = forms.URLField(required=False)
```
Comment (Coop56) [inline mapstory/apps/organizations/forms.py:33]: Do we only want three URLs as maximum for orgs to list?

File: mapstory/apps/organizations/models.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        return OrganizationMembership.objects.filter(organization=self, is_admin=True)
+
+    def remove_member(self, user):
+        """Removes a member's membership.
+
+        Deletes a Membership for the given user.
+
+        :param user: The user to be removed from the Organization
+        """
+        membership = OrganizationMembership.objects.get(user=user, organization=self)
```
Comment (Coop56) [inline mapstory/apps/organizations/models.py]: What happens here if they aren't a member?

File: mapstory/templates/_header.html [hunk 3]
```diff
@@ -59,13 +59,14 @@ <h1 class="site-title" style="background:url('{{ MEDIA_URL }}{{ site.assets.logo
...
                             <li class="divider"></li>
-                             <li><a href="{% url "messages_inbox" %}">{% trans "Messages" %}</a></li>
+                            <li><a href="{% url "messages_inbox" %}">{% trans "Messages" %}</a></li>
                             <li><a href="{% url 'edit_profile' user.username %}">{% trans "Edit Profile" %}</a></li>
                             <li><a href="{% url 'account_password' %}">{% trans "Change Password" %}</a></li>
                             <li><a href="{% url 'profile_detail' slug=user.username %}?showNotifications=true">{% trans "Notification Settings" %}</a></li>
                             <li class="divider"></li>
                             <li><a href="{% url 'getpage' 'started' %}">{% trans "Help" %}</a></li>
                             <li class="divider"></li>
+                        {#TODO: Add Organizations menu here#}
```
Comment (Coop56) [inline mapstory/templates/_header.html]: Did you mean to finish this TODO before the PR?

File: mapstory/apps/organizations/forms.py (hunk 1 above)
Comment (josellausas) [reply mapstory/apps/organizations/forms.py:33]: Created issue #1090 for this

**Review by Coop56 — APPROVED** [review]

**Pull request merged**